packages/notifications/push/amplify_push_notifications/ios/Classes/AmplifyPushNotificationsPlugin+AppDelegate.swift

```diff
@@ -0,0 +1,89 @@
+            EventsStreamHandlers.shared.tokenReceived.sendEvent(payload: ["token": token])
+        }
+    }
+
+    public func application(
+        _ application: UIApplication,
+        didFinishLaunchingWithOptions launchOptions: [AnyHashable : Any] = [:]
+    ) -> Bool {
+        if let remoteNotification = launchOptions[UIApplication.LaunchOptionsKey.remoteNotification] as? [AnyHashable: Any] {
+            // If the App is awaken by a remote notification when background mode is enabled
+            if application.applicationState == .background {
+                isBackgroundMode = true
+            }
+
+            // When the App is launched (from terminated state) by tapping on a notification
+            // This happens whne background mode is not enabled
+            if application.applicationState != .background {
+                launchNotification = remoteNotification
+            }
+        }
+
+        return true
+    }
+
+    public func application(
+        _ application: UIApplication,
+        didFailToRegisterForRemoteNotificationsWithError error: Error
+    ) {
+        print("registerForRemoteNotifications failed with error: \(error).")
+
+        EventsStreamHandlers.shared.tokenReceived.sendError(
+            error: FlutterError(code: "DeviceTokenError", message: error.localizedDescription, details: error)
+        )
+    }
+
+    public func application(
+        _ application: UIApplication,
+        didReceiveRemoteNotification userInfo: [AnyHashable : Any],
+        fetchCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void
+    ) -> Bool {
+        if UIApplication.shared.applicationState == .background {
+            let completionHandlerId = UUID().uuidString
+            var userInfoCopy = userInfo
+
+            remoteNotificationCompletionHandlers[completionHandlerId] = completionHandler
+            userInfoCopy[completionHandlerIdKey] = completionHandlerId
+
+            EventsStreamHandlers.shared.backgroundMessageReceived.sendEvent(payload: userInfoCopy)
+        } else if UIApplication.shared.applicationState == .inactive {
+            // When the App entering foreground by tapping on a notification
+            // and it's previously awaken by a remote notification, we treat the tapped notification
+            // as the launchNotitication
+            if (isBackgroundMode) {
+                isBackgroundMode = false
+                launchNotification = userInfo
+                EventsStreamHandlers.shared.launchNotificationOpened.sendEvent(payload: userInfo)
+            } else {
+                EventsStreamHandlers.shared.notificationOpened.sendEvent(payload: userInfo)
+            }
+        } else {
+            EventsStreamHandlers.shared.foregroundMessageReceived.sendEvent(payload: userInfo)
+        }
+
+        return true
+    }
+
+    public func applicationDidBecomeActive(_ application: UIApplication) {
+        application.registerForRemoteNotifications()
```

Comment:
Should this happen in `didFinishLaunchingWithOptions`?

Reply:
Same Q

Reply:
No we discussed during the development of react-native iOS module, that we invoke `registerForRemoteNotifications`  when App becomes active, according to Apple documentation `registerForRemoteNotifications` returns fast after the very first call there shouldn't be a performance concern here.